src/index.ts

```diff
@@ -184,6 +186,185 @@ export function HTPipeAttachData<
+  >
+>(
+  left: TLeft,
+  right: TRight
+): HasFinalAuthorize<TContextInLeft, TContextOutLeft>;
+
+// right has finalAuthorize and left does not
+export function HTPipeFinalAuthorize<
+  TLeft extends MightHaveFinalAuthorize<
+    any,
+    | boolean
+    | (TRight extends HasFinalAuthorize<any, any>
+        ? Pick<
+            Parameters<TRight['finalAuthorize']>[0],
+            keyof PromiseResolveOrSync<
+              ReturnType<
+                TLeft extends HasFinalAuthorize<any, any>
+                  ? TLeft['finalAuthorize']
+                  : () => {}
+              >
+            >
+          >
+        : any)
+  >,
+  TRight extends HasFinalAuthorize<any, any>,
+  TContextInRight extends Parameters<TRight['finalAuthorize']>[0],
+  TContextOutRight extends PromiseResolveOrSync<
+    ReturnType<TRight['finalAuthorize']>
+  >
+>(
+  left: TLeft,
+  right: TRight
+): HasFinalAuthorize<TContextInRight, TContextOutRight>;
+
+// right and left doesn't have preAuthorize
+export function HTPipeFinalAuthorize<
+  TLeft extends MightHaveFinalAuthorize<
+    any,
+    | boolean
+    | (TRight extends HasFinalAuthorize<any, any>
+        ? Pick<
+            Parameters<TRight['finalAuthorize']>[0],
+            keyof PromiseResolveOrSync<
+              ReturnType<
+                TLeft extends HasFinalAuthorize<any, any>
+                  ? TLeft['finalAuthorize']
+                  : () => {}
+              >
+            >
+          >
+        : any)
+  >,
+  TRight extends MightHaveFinalAuthorize<any, any>
+>(left: TLeft, right: TRight): {};
+
+// finalAuthorize main function
+export function HTPipeFinalAuthorize<
+  TLeft extends MightHaveFinalAuthorize<any, any>,
+  TRight extends MightHaveFinalAuthorize<any, any>,
+  TContextInLeft extends TLeft extends HasFinalAuthorize<any, any>
+    ? Parameters<TLeft['finalAuthorize']>[0]
+    : never,
+  TContextInRight extends TRight extends HasFinalAuthorize<any, any>
+    ? Parameters<TRight['finalAuthorize']>[0]
+    : never,
+  TContextOutLeft extends TLeft extends HasFinalAuthorize<any, any>
+    ? PromiseResolveOrSync<ReturnType<TLeft['finalAuthorize']>>
+    : never,
+  TContextOutRight extends TRight extends HasFinalAuthorize<any, any>
+    ? PromiseResolveOrSync<ReturnType<TRight['finalAuthorize']>>
+    : never
+>(left: TLeft, right: TRight) {
+  if (isHasFinalAuthorize(left) && isHasFinalAuthorize(right)) {
+    return {
+      finalAuthorize: async (
+        context: TContextOutLeft extends TContextInRight
+          ? TContextInLeft
+          : TContextInRight & TContextInLeft
+      ) => {
+        const leftOut =
+          (await Promise.resolve(left.finalAuthorize(context))) || false;
+        const leftPassed = authorizationPassed(leftOut);
+        if (!leftPassed) {
+          return false;
+        }
+        const leftContextOut = leftOut === true ? {} : leftOut;
+        const rightIn = {
+          ...context,
+          ...leftContextOut,
+        };
```

Comment:
- [ ] same as in the other one - just return rightOut

Reply:
done